src/Microsoft.Diagnostics.Monitoring.HostingStartup/ParameterCapturing/Eventing/AsyncParameterCapturingEventSource.cs

```diff
@@ -0,0 +1,125 @@
+            _eventSource = eventSource;
+
+            _thread = new(ThreadLoop)
+            {
+                Priority = ThreadPriority.BelowNormal,
+                IsBackground = true,
+                Name = BackgroundLoggingThreadName
+            };
+            _thread.Start();
+        }
+
+        public void Dispose()
+        {
+            _pendingEvents.CompleteAdding();
+            _cts.Cancel();
+
+            _thread.Join();
+
+            _cts.Dispose();
+            _pendingEvents.Dispose();
+        }
+
+        public bool IsEnabled => _eventSource.IsEnabled();
+
+        public void OnCapturedParameters(
+            Guid requestId,
+            string methodName,
+            string methodModuleName,
+            string? methodDeclaringTypeName,
+            ResolvedParameterInfo[] parameters
+            )
+        {
+            Activity? currentActivity = Activity.Current;
+
+            ScheduleAction((cancellationToken) =>
+            {
+                Guid captureId = Guid.NewGuid();
+
+                cancellationToken.ThrowIfCancellationRequested();
+
+                _eventSource.CapturedParameterStart(
+                    requestId,
+                    captureId,
+                    currentActivity?.Id ?? string.Empty,
+                    methodName,
+                    methodModuleName,
+                    methodDeclaringTypeName ?? string.Empty);
+
+                foreach (ResolvedParameterInfo param in parameters)
+                {
+                    cancellationToken.ThrowIfCancellationRequested();
+
+                    _eventSource.CapturedParameter(
+                        requestId,
+                        captureId,
+                        param.Name ?? string.Empty,
+                        param.Type ?? string.Empty,
+                        param.TypeModuleName ?? string.Empty,
+                        param.Value,
+                        param.Attributes,
+                        param.IsByRef);
+                }
+
+                cancellationToken.ThrowIfCancellationRequested();
+                _eventSource.CapturedParameterStop(requestId, captureId);
+            });
+        }
+
+        public void FailedToCapture(
+            Guid requestId,
+            ParameterCapturingEvents.CapturingFailedReason reason,
+            string details) => ScheduleAction(_ =>
+        {
+            _eventSource.FailedToCapture(requestId, reason, details);
+        });
+
+        private void ScheduleAction(Action<CancellationToken> action)
+        {
+            _ = _pendingEvents.TryAdd(action);
+        }
+
+        private void ThreadLoop()
+        {
+            using IDisposable _ = MonitorExecutionContextTracker.MonitorScope();
+            using IDisposable __ = new NoProbeScope();
+
+            try
+            {
+                while (_pendingEvents.TryTake(out Action<CancellationToken>? eventAction, Timeout.Infinite, _cts.Token))
```

Comment:
Extract the token first from _cts. Otherwise, disposing asynchronously will cause .Token to throw ObjectDisposedException instead of cancelling as expected.

Reply:
Actually nvm, you join this thread before disposing.